hrana/lib/src/rpc_http_client.dart

```diff
@@ -0,0 +1,267 @@
+          code: null,
+        ),
+    };
+  }
+
+  @override
+  Future<SqlStreamId> openStream() async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    final streamId = SqlStreamId(_nextStreamId++);
+    await _runPipeline(
+      streamId: streamId,
+      request: null,
+    );
+    return streamId;
+  }
+
+  @override
+  Future<void> closeStream(SqlStreamId id) async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    try {
+      final response = await _runPipeline(
+        streamId: id,
+        request: proto.StreamRequest(close: proto.CloseStreamReq()),
+      );
+      assert(response.whichResponse() == proto.StreamResponse_Response.close);
+    } finally {
+      _streams.remove(id);
+    }
+  }
+
+  @override
+  Future<StatementResult> executeStatement(
+    SqlStreamId stream,
+    StatementDescription stmt,
+  ) async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    final response = await _runPipeline(
+      streamId: stream,
+      request: proto.StreamRequest(
+        execute: proto.ExecuteStreamReq(stmt: stmt.toStatement()),
+      ),
+    );
+    assert(response.whichResponse() == proto.StreamResponse_Response.execute);
+    return StatementResult.fromProto(response.execute.result);
+  }
+
+  @override
+  Future<proto.BatchResult> runBatch(
+    SqlStreamId stream,
+    proto.Batch batch,
+  ) async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    final response = await _runPipeline(
+      streamId: stream,
+      request: proto.StreamRequest(batch: proto.BatchStreamReq(batch: batch)),
+    );
+    assert(response.whichResponse() == proto.StreamResponse_Response.batch);
+    return response.batch.result;
+  }
+
+  @override
+  Future<SqlTextId> storeSql(SqlStreamId stream, String sql) async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    final textId = SqlTextId(_nextSqlTextId++);
+    final response = await _runPipeline(
+      streamId: stream,
+      request: proto.StreamRequest(
+        storeSql: proto.StoreSqlReq(sql: sql, sqlId: textId.id),
+      ),
+    );
+    assert(response.whichResponse() == proto.StreamResponse_Response.storeSql);
+    return textId;
+  }
+
+  @override
+  Future<void> closeSql(SqlStreamId stream, SqlTextId id) async {
+    if (_isClosed) {
+      throw const ConnectionClosed();
+    }
+    final response = await _runPipeline(
+      streamId: stream,
+      request: proto.StreamRequest(
+        closeSql: proto.CloseSqlReq(sqlId: id.id),
+      ),
+    );
+    assert(response.whichResponse() == proto.StreamResponse_Response.closeSql);
+  }
+
+  var _isClosed = false;
+
+  @override
+  bool get isClosed => _isClosed;
+
+  @override
+  Future<void> get closed => _closed.future;
+
+  @override
+  Future<void> close() async {
+    if (_isClosed) {
+      return;
+    }
+    _isClosed = true;
+    _streams.clear();
+    _httpClient.close();
+    _closed.complete();
```

Comment:
Do you know whether we should attempt to close pending streams here to avoid leaking data on the server? Or are these short-lived enough to not cause issues?

Reply:
Originally, I had it closing them all, but then I realized that the `Database` class only ever handles one stream at a time and closes it in its `close` method. Added it back since it doesn't hurt 👍 

There is no risk to not closing them as they do expire pretty aggressively (10 seconds, I believe). In fact, this probably needs a keep-alive mechanism, but I wanted to get the initial code in first to test it out.